src/unix/fs.c

```diff
@@ -474,6 +474,72 @@ static ssize_t uv__fs_scandir(uv_fs_t* req) {
   return n;
 }
 
+static int uv__fs_opendir(uv_fs_t* req) {
+  req->dir = uv__malloc(sizeof(uv_dir_t));
+
+  if (req->dir == NULL)
+    return -1;
+
+  req->dir->dir = opendir(req->path);
+
+  if (req->dir->dir == NULL) {
+    uv__free(req->dir);
+    req->dir = NULL;
+    return -1;
+  }
+
+  return 0;
+}
+
+static int uv__fs_readdir(uv_fs_t* req) {
+  uv_dirent_t* dirent;
+  struct dirent* res;
+  unsigned int dirent_idx;
+
+  assert(req->dir != NULL);
+  assert(req->dir->dir != NULL);
+  assert(req->dirents != NULL);
+
+  for (dirent_idx = 0; dirent_idx < req->nentries; ++dirent_idx) {
+    /* readdir() returns NULL on end of directory, as well as on error. errno
+       is used to differentiate between the two conditions. */
+    errno = 0;
+    res = readdir(req->dir->dir);
```

Comment:
Hey @cjihrig, at least on *nix you will need a mutex per dir and lock it around readdir() because it is not thread safe. I noted that in the previous PR https://github.com/libuv/libuv/pull/416#issuecomment-381521263